File: 03-Session-Handling/app/src/main/java/com/auth0/sessiondemo/MainActivity.java
package com.auth0.sessiondemo;

import android.content.Intent;
import android.os.Bundle;
import android.support.v7.app.AppCompatActivity;
import android.view.View;
import android.widget.Button;
import android.widget.Toast;

import com.auth0.android.Auth0;
import com.auth0.android.authentication.AuthenticationAPIClient;
import com.auth0.android.authentication.AuthenticationException;
import com.auth0.android.callback.BaseCallback;
import com.auth0.android.result.Credentials;
import com.auth0.sessiondemo.utils.CredentialsManager;


public class MainActivity extends AppCompatActivity {

    private AuthenticationAPIClient mAuthenticationClient;

    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        setContentView(R.layout.activity_main);

        Auth0 auth0 = new Auth0(getString(R.string.auth0_client_id), getString(R.string.auth0_domain));
        auth0.setOIDCConformant(true);
        mAuthenticationClient = new AuthenticationAPIClient(auth0);

        Button refreshTokenButton = (Button) findViewById(R.id.refreshTokenButton);
        Button logoutButton = (Button) findViewById(R.id.logout);

        refreshTokenButton.setOnClickListener(new View.OnClickListener() {
            @Override
            public void onClick(View v) {
                renewAuthentication();
            }
        });
        logoutButton.setOnClickListener(new View.OnClickListener() {
            @Override
            public void onClick(View v) {
                logout();
            }
        });
    }

    private void renewAuthentication() {
        String refreshToken = CredentialsManager.getCredentials(this).getRefreshToken();
        mAuthenticationClient.renewAuth(refreshToken).start(new BaseCallback<Credentials, AuthenticationException>() {
            @Override
            public void onSuccess(final Credentials payload) {
                MainActivity.this.runOnUiThread(new Runnable() {
                    public void run() {
                        Toast.makeText(MainActivity.this, "New access_token: " + payload.getAccessToken(), Toast.LENGTH_SHORT).show();
                    }
                });
            }

            @Override
            public void onFailure(AuthenticationException error) {
                MainActivity.this.runOnUiThread(new Runnable() {
                    public void run() {
                        Toast.makeText(MainActivity.this, "Failed to get the new access_token", Toast.LENGTH_SHORT).show();
                    }
                });
            }
        });
    }


    private void logout() {
        CredentialsManager.deleteCredentials(this);
        startActivity(new Intent(this, LoginActivity.class));
        finish();
    }
}

File: app/src/main/java/com/auth0/samples/MainActivity.kt
package com.auth0.samples

import android.os.Bundle
import android.widget.Toast
import androidx.activity.*
import androidx.activity.compose.setContent
import androidx.compose.foundation.layout.*
import androidx.compose.material3.*
import androidx.compose.runtime.*
import androidx.compose.ui.Modifier
import androidx.compose.ui.unit.*
import com.auth0.android.Auth0
import com.auth0.android.authentication.*
import com.auth0.android.authentication.storage.*
import com.auth0.android.callback.Callback
import com.auth0.android.provider.WebAuthProvider
import com.auth0.android.result.Credentials

class MainActivity : ComponentActivity() {
    /* highlight-start account-setup */
    // Initialized lazily because getString() requires Activity context
    private val account: Auth0 by lazy {
        Auth0.getInstance(
            getString(R.string.com_auth0_client_id),
            getString(R.string.com_auth0_domain)
        )
    }
    private val manager: CredentialsManager by lazy {
        CredentialsManager(AuthenticationAPIClient(account), SharedPreferencesStorage(this))
    }

    // Compose re-renders the UI when these change
    private var credentials by mutableStateOf<Credentials?>(null)
    private var isLoading by mutableStateOf(true)
    /* highlight-end account-setup */

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        enableEdgeToEdge()

        /* highlight-start credentials-manager */
        // Restore session from disk; refreshes the access token if expired
        manager.getCredentials(object : Callback<Credentials, CredentialsManagerException> {
            override fun onSuccess(result: Credentials) {
                credentials = result
                isLoading = false
            }

            override fun onFailure(error: CredentialsManagerException) {
                isLoading = false
            }
        })
        /* highlight-end credentials-manager */

        setContent {
            Column(
                modifier = Modifier
                    .fillMaxSize()
                    .systemBarsPadding()
                    .padding(horizontal = 16.dp),
                verticalArrangement = Arrangement.spacedBy(16.dp)
            ) {
                /* highlight-start app */
                if (isLoading) {
                    Text("Loading...")
                } else if (credentials == null) {
                    Button(onClick = { login("signup") }) { Text("Sign Up") }
                    Button(onClick = { login() }) { Text("Log In") }
                } else {
                    Text("Logged in as ${credentials?.user?.email}", fontSize = 18.sp)
                    Column {
                        Text("sub: ${credentials?.user?.getId()}")
                        Text("name: ${credentials?.user?.name}")
                        Text("email: ${credentials?.user?.email}")
                        Text("email_verified: ${credentials?.user?.isEmailVerified}")
                        Text("nickname: ${credentials?.user?.nickname}")
                        Text("picture: ${credentials?.user?.pictureURL}")
                    }
                    Button(onClick = { logout() }) { Text("Log Out") }
                }
                /* highlight-end app */
            }
        }
    }

    /* highlight-start login */
    private fun login(screenHint: String? = null) {
        WebAuthProvider.login(account)
            .withScheme(getString(R.string.com_auth0_scheme))
            // offline_access: requests a refresh token for session persistence
            .withScope("openid profile email offline_access")
            .withParameters(buildMap { screenHint?.let { put("screen_hint", it) } })
            .start(this, object : Callback<Credentials, AuthenticationException> {
                override fun onSuccess(result: Credentials) {
                    credentials = result
                    manager.saveCredentials(result)
                }

                override fun onFailure(error: AuthenticationException) {
                    Toast.makeText(this@MainActivity, error.message, Toast.LENGTH_SHORT).show()
                }
            })
    }
    /* highlight-end login */

    /* highlight-start logout */
    private fun logout() {
        WebAuthProvider.logout(account)
            .withScheme(getString(R.string.com_auth0_scheme))
            .start(this, object : Callback<Void?, AuthenticationException> {
                override fun onSuccess(result: Void?) {
                    credentials = null
                    manager.clearCredentials()
                }

                override fun onFailure(error: AuthenticationException) {
                    Toast.makeText(this@MainActivity, error.message, Toast.LENGTH_SHORT).show()
                }
            })
    }
    /* highlight-end logout */
}

